src/executables/master_server_monitor.hpp

```diff
@@ -26,58 +26,143 @@
+    };
+
     class server_entry
     {
       public:
-	server_entry (int pid, const char *exec_path, char *args, CSS_CONN_ENTRY *conn);
+	server_entry (int pid, std::string exec_path, std::string args,
+		      std::chrono::steady_clock::time_point revive_time);
 	~server_entry () {};
 
 	server_entry (const server_entry &) = delete;
-	server_entry (server_entry &&) = delete;
+	server_entry (server_entry &&) = default;
 
 	server_entry &operator= (const server_entry &) = delete;
-	server_entry &operator= (server_entry &&) = default;
+	server_entry &operator= (server_entry &&other)
+	{
+	  if (this != &other)
+	    {
+	      m_pid = other.m_pid;
+	      m_need_revive = other.m_need_revive;
+	      m_last_revive_time = other.m_last_revive_time;
+	      m_exec_path = other.m_exec_path;
+	      m_argv = other.m_argv;
+	    }
+	  return *this;
+	}
+
+	int get_pid () const;
+	std::string get_exec_path () const;
+	std::vector<std::string> get_argv () const;
+	bool get_need_revive () const;
+	std::chrono::steady_clock::time_point get_last_revive_time () const;
 
-	CSS_CONN_ENTRY *get_conn () const;
+	void set_pid (int pid);
+	void set_exec_path (std::string exec_path);
+	void set_need_revive (bool need_revive);
+	void set_last_revive_time (std::chrono::steady_clock::time_point revive_time);
+
+	void proc_make_arg (const std::string &args);
 
       private:
-	void proc_make_arg (char *args);
-
-	int m_pid;                                    // process ID of server process
-	std::string m_exec_path;                      // executable path of server process
-	std::vector<std::string> m_argv;              // arguments of server process
-	CSS_CONN_ENTRY *m_conn;                       // connection entry of server process
-	timeval m_last_revive_time;                   // latest revive time
-	bool m_need_revive;                           // need to revive (true if the server is abnormally terminated)
+	int m_pid;                                                 // process ID of server process
+	std::string m_exec_path;                                   // executable path of server process
+	std::vector<std::string> m_argv;                           // arguments of server process
+	volatile bool m_need_revive;                               // need to be revived by monitoring thread
+	std::chrono::steady_clock::time_point m_last_revive_time;  // last revive time
     };
 
-    server_monitor ();
-    ~server_monitor ();
+    std::unordered_map <std::string, server_entry> m_server_entry_map;  // map of server entries
 
-    server_monitor (const server_monitor &) = delete;
-    server_monitor (server_monitor &&) = delete;
+    std::unique_ptr<std::thread> m_monitoring_thread;                   // monitoring thread
+    std::queue<job> m_job_queue;                                        // job queue for monitoring thread
+    volatile bool m_thread_shutdown;                                    // flag to shutdown monitoring thread
```

Comment:
m_server_monitor_mutex에 의해 보호되기 때문에 volatile은 불필요 합니다.

```suggestion
    bool m_thread_shutdown;                                    // flag to shutdown monitoring thread
```

Reply:
수정하였습니다.